filters/openpolicyagent/openpolicyagent.go

```diff
@@ -0,0 +1,325 @@
+	if err != nil {
+		return nil, err
+	}
+
+	discovery, err := discovery.New(manager, discovery.Factories(map[string]plugins.Factory{envoy.PluginName: envoy.Factory{}}))
+	if err != nil {
+		return nil, err
+	}
+
+	manager.Register("discovery", discovery)
+
+	opa := &OpenPolicyAgentInstance{
+		instanceConfig: instanceConfig,
+		manager:        manager,
+		opaConfig:      opaConfig,
+		bundleName:     bundleName,
+
+		preparedQueryDoOnce:    new(sync.Once),
+		interQueryBuiltinCache: iCache.NewInterQueryCache(manager.InterQueryBuiltinCacheConfig()),
+	}
+
+	manager.RegisterCompilerTrigger(opa.compilerUpdated)
+
+	return opa, nil
+}
+
+// Start asynchronously starts the policy engine's plugins that download
+// policies, report status, etc.
+func (opa *OpenPolicyAgentInstance) Start(ctx context.Context) error {
+	return opa.manager.Start(ctx)
+}
+
+func (opa *OpenPolicyAgentInstance) waitPluginsReady(checkInterval, timeout time.Duration) error {
+	if timeout <= 0 {
+		return nil
+	}
+
+	// check readiness of all plugins
+	pluginsReady := func() bool {
+		for _, status := range opa.manager.PluginStatus() {
+			if status != nil && status.State != plugins.StateOK {
+				return false
+			}
+		}
+		return true
+	}
+
+	opa.Logger().Debug("Waiting for plugins activation (%v).", timeout)
+
+	return opautil.WaitFunc(pluginsReady, checkInterval, timeout)
+}
+
+func (opa *OpenPolicyAgentInstance) InstanceConfig() *OpenPolicyAgentInstanceConfig {
+	return &opa.instanceConfig
+}
+
+func (opa *OpenPolicyAgentInstance) compilerUpdated(txn storage.Transaction) {
+	opa.preparedQueryDoOnce = new(sync.Once)
+}
+
+func (opa *OpenPolicyAgentInstance) EnvoyPluginConfig() envoy.PluginConfig {
+	if plugin, ok := opa.manager.Plugin(envoy.PluginName).(*envoy.Plugin); ok {
+		return plugin.GetConfig()
+	}
+
+	defaultConfig := envoy.PluginConfig{
+		Path:   "envoy/authz/allow",
+		DryRun: false,
+	}
+	defaultConfig.ParseQuery()
+	return defaultConfig
+}
+
+func (opa *OpenPolicyAgentInstance) StartSpanFromContext(ctx context.Context) (opentracing.Span, context.Context) {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "open-policy-agent")
+	span.SetTag("bundle_name", opa.bundleName)
+	return span, ctx
+}
+
+func (opa *OpenPolicyAgentInstance) MetricsKey(key string) string {
+	return key + "." + opa.bundleName
+}
+
+func (opa *OpenPolicyAgentInstance) ParsedQuery() ast.Body {
+	return opa.EnvoyPluginConfig().ParsedQuery
+}
+
+func (opa *OpenPolicyAgentInstance) Store() storage.Store            { return opa.manager.Store }
+func (opa *OpenPolicyAgentInstance) Compiler() *ast.Compiler         { return opa.manager.GetCompiler() }
+func (opa *OpenPolicyAgentInstance) Runtime() *ast.Term              { return opa.manager.Info }
+func (opa *OpenPolicyAgentInstance) Logger() logging.Logger          { return opa.manager.Logger() }
+func (opa *OpenPolicyAgentInstance) PreparedQueryDoOnce() *sync.Once { return opa.preparedQueryDoOnce }
```

Comment:
I wonder who is calling this method. I don't find it in the PR code.

Reply:
These methods implement the envoyauth.EvalContext interface that is passed to the Eval method. I can also create an extra struct or am open to other ideas to make it more readable